### PR TITLE
Prevent translations from being loaded too early

### PR DESCRIPTION
**Pull request by swissspidy**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Since WordPress 4.6, WordPress has supported so-called just-in-time translation loading. Whenever it encounters a `__()` call, it tries to load the translations for that domain if it hasn't already. That means one theoretically doesn't need `load_plugin_textdomain()` anymore, though it still can be used.

Ideally, this should only happen after the `init` hook, as that's when the current user has been fully set up in WordPress. This is important as users can set their locale in the user profile, which is used in the admin. Triggering translation loading before that means they will be loaded in the wrong locale. Unfortunately, some plugins are `_doing_it_wrong()` and load translations right when the plugin is executed, or upon `plugins_loaded`.

In [my attempt to identify such plugins](https://github.com/WordPress/wordpress-develop/pull/6462), I noticed that WooCommerce was a particularly stubborn case. There is _a lot_ of code triggering just-in-time translation loading too early. This is particularly suboptimal because WooCommerce itself calls `load_plugin_textdomain()` later on on `init`, causing yet another translation call, which is just unnecessary at this point and not good for performance.

This PR is an attempt at identifying and eliminating all cases in WooCommerce triggering just-in-time translation loading before `init`.

This way, if core starts [adding warnings for such cases](https://github.com/WordPress/wordpress-develop/pull/6462), there won't be any warnings anymore for WooCommerce.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Some examples:

**Remove autoloader**

1. Set site language to German (de_DE)
2. Set user language to French (fr_FR)
3. Remove `vendor/autoload_packages.php` file so the missing autoloader warning kicks in
4. Visit dashboard
5. Without this PR: See autoloader admin notice in fr_FR, but error log entry in de_DE
6. With this PR: the error log entry is in en_US, as I removed the translation call

**Temporarily remove `load_plugin_textdomain()` call**

1. Set site language to German (de_DE)
2. Set user language to French (fr_FR)
3. Temporarily do an early return in `\WooCommerce::load_plugin_textdomain()` so that `load_plugin_textdomain()` isn't called
4. Visit some WooCommerce admin screen
5. Without this PR: Everything is in en_US or de_DE
7. With this PR: All strings are in fr_FR

**With core PR**

1. Apply https://github.com/WordPress/wordpress-develop/pull/6462
2. Activate WooCommerce
3. See `_doing_it_wrong()` warnings

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.
Adjust code loading order to prevent just-in-time translation loading from happening too early.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by ralucaStan** [comment]
@swissspidy thank you for taking the time to work on this PR. There are some conflicts on your PR as there were some changes on the Additional fields side. Would you mind fixing those and updating the branch?

**Comment by swissspidy** [comment]
Sure, addressed now! 👍

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php
Comment (coreymckrill) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: 🤔 Does this one really need to change? Looks like it doesn't actually use any `gettext` functions in the callback, just queues up a callback on another action hook that's after `init`...

File: plugins/woocommerce/includes/class-woocommerce.php
Comment (coreymckrill) [inline plugins/woocommerce/includes/class-woocommerce.php]: From [the docs](https://github.com/Automattic/jetpack/tree/391f1f22abbc8ba9c34344d14e659ecaf92df936/projects/packages/config#usage), it looks like this one needs to be called "at or before plugins_loaded". So we can't change this to `init`. Does this mean the "WooCommerce" string in the linked callback shouldn't be translated?

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php
Comment (coreymckrill) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php]: I see why this is added, so that `fields_locations` can be populated in the constructor, but it's obviously not ideal since it means the keys have to be updated in two places if they change. But, since both the `core_fields` and the `fields_locations` properties in this class are set to private, we could do the same treatment for `fields_locations`, i.e. create a new private method called `get_fields_locations` and use that instead of referencing the class property. Then you could remove the property assignment from the class constructor.

File: plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: Like in the WC_Helper_Updater class, is this change necessary? In the callback, it looks like all we're doing is adding callbacks to some other hooks that fire after init.

**Review by coreymckrill — COMMENTED** [review]
@swissspidy Thanks for this PR! I'm excited that this could potentially be a performance improvement for WooCommerce, and also improve how we implement translations.

I'm going to take another pass at this to do some specific testing (and others may as well) but I wanted to get some initial thoughts out here.

Switching a bunch of things from one action hook (`plugins_loaded`) to a later hook (`init`) makes me nervous because there are so many other plugins that integrate with WooCommerce, we can't really anticipate what assumptions they might be making about when things will be available, or what will break if we make a change like this. That said, it looks like several of these changes you make are in our `Internal` namespace, and so are probably relatively safe. I did note a couple that seem unnecessary because they don't look like they are actually triggering any gettext calls, but I could be wrong about that.

> This is particularly suboptimal because WooCommerce itself calls load_plugin_textdomain() later on on init, causing yet another translation call, which is just unnecessary at this point and not good for performance.

This is a good call out, and it's something I had actually noticed just recently, and was wondering if we still needed it. I notice that you didn't remove it as part of this PR though. Do you think that would be worth doing here?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php
Comment (swissspidy) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: You're right, it's not. I was getting fatal errors from `WC_Woo_Update_Manager_Plugin::is_plugin_active()` though because `is_plugin_active_for_network` wasn't yet available. That can be addressed better.

File: plugins/woocommerce/includes/class-woocommerce.php
Comment (swissspidy) [reply plugins/woocommerce/includes/class-woocommerce.php]: Either that, or that config package is updated to work on `init` instead or receive a callback or something (so that the translation function isn't immediately invoked)

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php
Comment (swissspidy) [reply plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php]: Agreed, the duplication is not ideal. A slight difference here is that `fields_locations` can be mutated via `register_checkout_field()`/`deregister_checkout_field`. Plus, it's accessed in quite a few places throughout the class. Any suggestions on how best to handle that?

File: plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php
Comment (swissspidy) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: This change is needed because `WPConsentAPI` is called by `OrderAttributionController`.

`OrderAttributionController` did things like accessing feature flags in `plugins_loaded`, which triggered translation loading. So I changed that to `init`. But that also means `WPConsentAPI` cannot register on `plugins_loaded` anymore, because at that point the hook has already fired.

**Comment by swissspidy** [comment]
> This is a good call out, and it's something I had actually noticed just recently, and was wondering if we still needed it. I notice that you didn't remove it as part of this PR though. Do you think that would be worth doing here?

I lack a bit of context on why that call exists there.

https://github.com/swissspidy/woocommerce/blob/7d83ad0e4ddd4646aed860f694dd3964acd06da6/plugins/woocommerce/includes/class-woocommerce.php#L805-L807

1. Why is `unload_textdomain()` being used? It seems like a hacky workaround to combat the early just-in-time translation loading, which this PR addresses.
2. `load_textdomain()` is reasonable, you want to support custom translations in the custom `wp-content/languages/woocommerce` directory.
3. `load_plugin_textdomain() - this call seems useless and only there for backward compatibility, if someone for some reason has translations in the `i18n/languages` folder. But they will be removed during upgrades anyway, soo... 🤷 
	* Note, if this gets removed, can remove the whole `i18n/languages` folder and the `Domain Path` header as well.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php
Comment (swissspidy) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: I changed this in 2fad07c2989be92be2b8c93e4d19cc00c28b876a, but not happy with it.

Actually, it would be even better to change this to `admin_init`, because the hooks there are only relevant for the admin, as they add admin notices.

Comment (swissspidy) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: Changed in c5c3e262c86defcd212f725e5c44674ebfe25279

File: plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php
Comment (rawdreeg) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: cc: @layoutd

Comment (layoutd) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: I just took it for a spin and this seems to play nicely with Order Attribution and Complianz with WP Consent API.

> [!WARNING] 
> HOWEVER, I was unable to test with block checkout because on this branch, the block checkout is not present:
> image


And if you swap branches and it was already present, this error shows:
image

On `trunk` both cases work as expected (as did Classic Checkout, which doesn't appear here either). Not sure if this is reproducible on other installs, I saw it with TT4 and Storefront.

Comment (layoutd) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: I retested on a fresh install because it struck me as odd, thinking maybe it was a plugin conflict. But I get the same error. For what it's worth, the **Install Checkout** button gives this:
image
image

And the documentation is on this page: https://fluidcheckout.com/docs/getting-started-fluid-checkout/#shortcode-vs-blocks

Comment (swissspidy) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: Okay so I suppose one of the action hook changes causes this to not work.

If someone can tell me how those are registered I can further look into it.

> For what it's worth, the Install Checkout button gives this:

What should the screenshots tell me? 🤔

Comment (layoutd) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: Screenshots: I was just wondering why a third-party product was being recommended.

Looking at the changes in the PR, it's this one in `src/Packages.php` that causes this issue: https://github.com/woocommerce/woocommerce/pull/47113/files#diff-c3a93b21b4c564d9c866729f93fcaf38838c2a67ab30e5cf8a0bd4db7e269fdc

Order Attribution / WP Consent API integration seem unaffected by the PR.

Comment (senadir) [reply plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php:34]: Same, I'm unable to test. However, changing the hook to run a bit early (priority 5) seems to fix it.

**Review by senadir — CHANGES_REQUESTED** [review]
I did give this a test as well, the switch to `plugins_loaded` broke Checkout, keeping it `init` and moving priority to 5 seems to partially fix it, but integration code isn't loading due to that, including built in payment methods. Some admin styles are loaded correctly as well, like the payment methods admin page.

**Comment by swissspidy** [comment]
> I did give this a test as well, the switch to `plugins_loaded` broke Checkout, keeping it `init` and moving priority to 5 seems to partially fix it, but integration code isn't loading due to that, including built in payment methods. 

Sorry, which switch are you referring to specifically? Can you comment inline on the changed line? In general, this PR is moving _away_ from `plugins_loaded` because that's fired too early. But your comment makes it sound like it's the other way around 🤔

**Comment by senadir** [comment]
`add_action( 'plugins_loaded', array( __CLASS__, 'on_init' ) );`: All works fine.
`add_action( 'init', array( __CLASS__, 'on_init' ) );`: Nothing blocks related loads.
`add_action( 'plugins_loaded', array( __CLASS__, 'on_init' ), 5 );`: Blocks load fine but integration code (like the built-in payment methods) don't load.

So I think it's worth digging deeper into integration code to see what's not loading.

**Comment by senadir** [comment]
Changing this line to `6` or anything after the initial one seems to work

https://github.com/woocommerce/woocommerce//blob/2821bc2a8c363a26c45fb3c149b551cd66ee3adb/plugins/woocommerce/src/Blocks/Payments/Api.php#L46

**Comment by swissspidy** [comment]
> `add_action( 'plugins_loaded', array( __CLASS__, 'on_init' ) );`: All works fine. `add_action( 'init', array( __CLASS__, 'on_init' ) );`: Nothing blocks related loads. `add_action( 'plugins_loaded', array( __CLASS__, 'on_init' ), 5 );`: Blocks load fine but integration code (like the built-in payment methods) don't load.
> 
> So I think it's worth digging deeper into integration code to see what's not loading.

Are you talking about `plugins/woocommerce/src/Internal/Integrations/WPConsentAPI.php` or `plugins/woocommerce/src/Packages.php`?? 🤔  There are many classes with such an `on_init` method.

**Comment by senadir** [comment]
My bad, it's `plugins/woocommerce/src/Packages.php`.

**Comment by senadir** [comment]
I did a round of testing and it looks like blocks are loading fine, I tested with several integrations as well and everything is fine. What's left is figuring out why button styling is missing on admin pages

This PR:
image

Trunk:
image

**Comment by swissspidy** [comment]
I think I resolved the `Packages` issue without re-introducing the translation issue.

The stack trace is basically this:

`Packages` -> `\Automattic\WooCommerce\Blocks\Package` -> `\Automattic\WooCommerce\Blocks\Domain\Bootstrap` -> `ShippingController` -> `LocalPickupUtils::is_local_pickup_enabled()` -> `__()`

The current workaround works, but makes it easy to call too early translation calls elsehwere, because everything in `Packages` runs so early.

----

The `init_jetpack_connection_config` case I "fixed" by not calling a translation function. Not great, but a better fix requires a change in that package.

**Comment by swissspidy** [comment]
@coreymckrill @senadir can someone help approve the workflow runs here?

**Comment by senadir** [comment]
Closing and reopening to force them to start.

**Comment by swissspidy** [comment]
They can't start without approval:

Screenshot 2024-06-03 at 16 13 35

**Comment by senadir** [comment]
I did approve them before closing and opening, just did that again.

**Review by senadir — APPROVED** [review]
I reviewed this yesterday and everything seems to be working fine in Checkout flows, what's left is more testing with some plugins and figuring out why tests aren't passing.

**Comment by swissspidy** [comment]
Tests should be passing now, but they need to be approved again.

**Comment by senadir** [comment]
FWIW `Build Live Branch / build (pull_request)` will continue to fail for now until we fix it in trunk (it has an issue with external authors).

**Comment by senadir** [comment]
I took anther look at the PR code and it's honestly not touching any critical stuff, the main stuff I can think about are the blocks related code (which I tested) + order attribution (with @layoutd tested).

I think this is good to go but I'd love for us to avoid reintroducing this in the future. @swissspidy is it possible for us to add a unit test that would check if translation is loaded too early and will fail if so? If you're ensure how to write the test, just an example of how that hook would look like is enough for me.

**Comment by swissspidy** [comment]
Discussed this with @senadir in person at WCEU:

I am planning to add warnings for translations being loaded too early in WP 6.7, see https://github.com/WordPress/wordpress-develop/pull/6462 and https://core.trac.wordpress.org/ticket/44937. This will make it easy to detect regressions in the future. Until then there is not a very good way to test this.

So my suggestion would be to merge this for now, accepting the small risk of a regression between now and WP 6.7.

**Comment by senadir** [comment]
Hey @swissspidy I discussed with internally with the release team and while I feel confident we can merge this, merging this today means it will land on WooCommerce 9.1 and won't get a lot of testing time, so I'd wait and merge it tomorrow so it lands on 9.2 and gives it even more time in local developer devs and testing time.

**Pull request merged**
